ProjektGopher/whisky
# Enable running in subdirectories

This is a proof of concept for being able to run hooks from a non-project-root directory within the git repository.

Resolves #75.

My testing setup is as follows:

* `cd .. && mkdir whisky-test && cd whisky-test`
  (`whisky-test` and `whisky` are now next to eachother)
* `git init`
* `mkdir subdirectory && cd subdirectory`
* `composer install`
* Add this to `composer.json`:
  ```
      "repositories": [
        {
            "type": "path",
            "url": "../../whisky",
            "options": {
                "symlink": true
            }
        }
    ],
  ```
* `composer update`
* `vendor/bin/whisky install`
* Update `whisky.json` to have the pre-commit hooks contain `"echo 'Running pre-commit hook'"`.

After making changes in the `whisky` source I needed to run `composer build` to update the symlinked binary so that I could test with it locally.

The biggest area that would need improvement that I can think of right now is what the story is for changing subdirectories. On install the subdirectory is now included in the pre-commit script as we need to change working directories (see `Hook.php:134`). Say the directory changes, there is no way or documentation on how to update Whisky's configuration. I suppose checking for existing lines in the `pre-commit` file containing `vendor/bin/whisky` and updating them would be the solution to this.

Please let me know if there's anything I missed or if you would like me to flesh out certain aspects.
